opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py

```diff
@@ -406,25 +445,55 @@ def use_span(
     ) -> Iterator[trace_api.Span]:
         """See `opentelemetry.trace.Tracer.use_span`."""
         try:
-            span_snapshot = self._current_span_slot.get()
-            self._current_span_slot.set(span)
+            span_snapshot = self.source.get_current_span()
+            self.source._current_span_slot.set(  # pylint:disable=protected-access
+                span
+            )
             try:
                 yield span
             finally:
-                self._current_span_slot.set(span_snapshot)
+                self.source._current_span_slot.set(  # pylint:disable=protected-access
+                    span_snapshot
+                )
         finally:
             if end_on_exit:
                 span.end()
 
+
+class TracerSource(trace_api.TracerSource):
+    def __init__(
+        self, sampler: sampling.Sampler = trace_api.sampling.ALWAYS_ON,
+    ):
+        # TODO: How should multiple TracerSources behave? Should they get their own contexts?
+        # This could be done by adding `str(id(self))` to the slot name.
+        self._current_span_slot = Context.register_slot("current_span")
+        self._active_span_processor = MultiSpanProcessor()
+        self.sampler = sampler
+
+    def get_tracer(
+        self,
+        instrumenting_library_name: str,
+        instrumenting_library_version: str = "",
+    ) -> "trace_api.Tracer":
+        if not instrumenting_library_name:  # Reject empty strings too.
+            instrumenting_library_name = "ERROR:MISSING LIB NAME"
+            logger.error("get_tracer called with missing library name.")
```

Comment:
Looking at the spec, it looks like the name should be optional? https://github.com/open-telemetry/opentelemetry-specification/pull/354/files#diff-ea4f4a46fe8755cf03f9fb3a6434cb0cR98

Reply:
No, it's actually required right now: https://github.com/open-telemetry/opentelemetry-specification/blob/a6a22409a2d242a6af097d72fd1241dd64e97219/specification/api-tracing.md#obtaining-a-tracer

Reply:
Ah. Thank you

Reply:
I think we ought to push back on this, the most common case is likely to be the default tracer, which doesn't need a name.

Reply:
Why do you think that? I think the most common case will be tracers created by instrumentations, which should definitely tell their name to OpenTelemetry.

Reply:
Even when you have no dedicated instrumentation library, you can use the name and version of your application here.

Reply:
(sharing an extra opinion) I don't think it's a big deal. two extra arguments of boilerplate isn't the best, but it's not a usability blocker.

I personally imagine tracing will be transparent to most people as the integrations will take care of injecting traces into the right locations. If there are enough extensions, the need for custom traces will probably be fairly minimal.